File: 0022-generate-parentheses/0022-generate-parentheses.dart
class Solution {
  List<String> generateParenthesis(int n) {
    List<String> res = [];
    solve(res, n, 0, 0, "");
    return res;
  }
  void solve(List<String> res, int n, int l, int r, String s){
    if(n == l && n == r){
        res.add(s);
        return;
    }
    if(n > l)
        solve(res, n, l+1, r, s+"(");
    if(l > r)
        solve(res, n, l, r+1, s+")");
  }
}

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        backtrack(res, n, 0, "", 0, 0);
        return res;
    }
    private void backtrack(List<String> res, int n, int count, String s, int openCount, int closeCount){
        if(openCount < 0 || openCount > n || openCount < closeCount) return;
        if(count/2 == n){
            res.add(s);
            return;
        }
        count++;
        backtrack(res, n, count, s + "(", openCount+1, closeCount);
        backtrack(res, n, count, s + ")", openCount, closeCount+1);
    }
}